### ISSUE TITLE
Null-aware elements

### ISSUE THREAD

**Issue by alorenzen** [dart-lang/language#323]
_**Admin comment**:_

This is being implementated: [feature specification](https://github.com/dart-lang/language/blob/main/accepted/future-releases/0323-null-aware-elements/feature-specification.md), [implementation issue](https://github.com/dart-lang/sdk/issues/55949), experiment flag `null-aware-elements`.

Null-aware elements offer a shorthand for adding an element to a list, if the element is non-null. Consider this code before the feature:

```dart
Stack(
  fit: StackFit.expand,
  children: [
    const AbsorbPointer(),
    if (widget.child != null) widget.child!,
  ],
)

```

The causes repetition of `widget.child`, and is verbose. With null-aware elements this becomes:

```dart
Stack(
  fit: StackFit.expand,
  children: [
    const AbsorbPointer(),
    ?widget.child,
  ],
)

```

---

**_Original issue contents:_**

I would love to have a "null-aware" operator for adding an element to a list if it is non-null.

```
[
  foo,
  ?bar,
]
```
This would be equivalent to:
```
[
  foo,
  if (bar != null) bar,
]
```

I've been trying out the new "UI-as-code" features in the angular codebase, and find that I can't convert a common pattern to use the new element syntax.

We often iterate over a list, adding elements to a new list when the value is non-null.

```
var result = [];
for (var node in nodes) {
  var value = convert(node);
  if (value != null) result.add(value);
}
```

It would be pretty nice if I could instead use the new syntax:
```
var result = [
  for (var node in nodes) ?convert(node);
]
```

**Comment by bwilkerson** [comment]
In the short term, you can nest the `if` inside the `for`:
```
var result = [
  for (var node in nodes) if (node != null) convert(node)
]
```

**Comment by alorenzen** [comment]
@bwilkerson I want to check the value of `convert(node)`, not `node` for nullability. You're example would require me to call `convert` twice, which I do not want to do:
```
var result = [
  for (var node in nodes) if (convert(node) != null) convert(node)
]
```

**Comment by leafpetersen** [comment]
cc @munificent

**Comment by munificent** [comment]
This is an interesting corner. We could do some kind of single-value null-aware element syntax like you suggest. I'm very hesitant to pile more semantics onto `?` because it's already quite overloaded in the grammar (postfix for nullable types, in null-aware operators, and in conditional expressions).

In many cases, you can use the other UI-as-code features like Brian suggests, but in this case you want to avoid any redundant computation. Of course, the natural way to do that is with a local variable, which makes me think building the list imperatively like you do today probably *is* the best approach.

**Comment by lrhn** [comment]
We could introduce a `let` construct.
```dart
[ let tmp = something in if (tmp != null) tmp ]
```
which is obviously not as short as `[? something]`, but it is more generally useful.
The current existing workaround is 
```dart
[ for (var tmp in [something]) if (tmp != null) tmp ]
```
or
```dart
[... [something].where(isNotNull)]
```
or (after NNBD)
```dart
[... [something].whereType<Object>()]
```
All of these will likely introduce an intermediate list, with only the first example being likely to have that optimized away.

**Comment by tatumizer** [comment]
See also #219.

**Comment by munificent** [comment]
I think the only difference is that in #219, you propose allowing `?` to elide arguments in argument lists as well. Otherwise, I think they're the same.

**Comment by tatumizer** [comment]
I was just trying to find more use cases for an operator, and the discussion mentioned even more potential uses, but this has to be filtered, for sure. Maybe there's a way to merge these 2 discussions?

**Comment by lrhn** [comment]
Just saw code of the form:
```dart
List<Property<Foo>> get elements => [
        something,
        other,
        orWhatnot,
        andYetOne,
        andFinally,
      ].whereNotNull().toList();
```
I tried to suggest using a literal with `[if (something != null) something, ...]`, to avoid creating two lists where only one is needed, but the names were fields, so promotion wouldn't work. With null-aware elements, it would literally just be:
```dart
List<Property<Foo>> get elements => [
        ?something,
        ?other,
        ?orWhatnot,
        ?andYetOne,
        ?andFinally,
      ];
```
That would be awesome!

**Comment by srawlins** [comment]
Agreed! Null-aware spread has spoiled me. We have a fancy null-aware element, but not a simple one.

**Comment by eernstg** [comment]
That's a neat trick! :smiley:

But I'm afraid it requires _every_ type to implement `iterator` (and a bunch of other methods), such that the null object can be the empty iterable, and every other object can be a one-element iterable, and that would not work. For instance, this creates a conflict for an object which is already an iterable yielding a bunch of other objects.

**Comment by lrhn** [comment]
If we had `Nullable<T>` as a type, shorthanded as `T?`, then we could assign methods to that. (C# has such a type, although it's probably a very special system type).

Or if we could iterate based on an extension `get iterator`, then we could assign one to `T?` only. Might surprise someone when their `Iterable<X>? f; something; [...f]` will give them a list containing a single iterable, not spread the iterable itself.

**Comment by PeterMcKinnis** [comment]
> But I'm afraid it requires every type to implement iterator (and a bunch of other methods), such that the null object can be the empty iterable, and every other object can be a one-element iterable, and that would not work. For instance, this creates a conflict for an object which is already an iterable yielding a bunch of other objects.

Interesting!  Didn't realize that nullables were implemented without a wrapper object.  I have always wondered why e.g. `int??` isn't a type.  Now I know.    Sigh...I wish we had first class discriminated unions.

**Comment by munificent** [comment]
In case you're curious, I wrote a blog post a while back about the rationale behind that design choice: https://medium.com/dartlang/why-nullable-types-7dd93c28c87a

**Comment by rakudrama** [comment]
> The current existing workaround is
>
>     [ for (var tmp in [something]) if (tmp != null) tmp ]
...
> All of these will likely introduce an intermediate list, with only the first example being likely to have that optimized away.

Some time ago I took a shot at getting dart2js to optimize using a singleton list as a 'let'-expression and it was surprisingly difficult to get anything reasonable. I did manage to remove the allocation, but there were all kinds of weird artifacts in the generated that would not exist with a 'let'-expression. I would say that "being likely to have that optimized away" is too optimistic.

**Comment by Reprevise** [comment]
This should work for maps too, like:

```dart
final String? seasonId = '1';
final queryParameters = <String, Object>{
  'seasonId': ?seasonId, // the '?' could instead go in front of the key, instead of the value
};
```

If Dart allowed for inferring the key, it would look better:

```dart
final queryParameters = <String, Object>{
  ?seasonId,
};
```

Then again, now that I'm reading it, maybe Dart shouldn't allow for inferring the key, looks weird 😅.

**Comment by Mike278** [comment]
FWIW patterns help with this - the example in the original post could be written as:

```dart

var result = [
  for (var node in nodes) 
    if (convert(node) case var converted?) 
      converted
];

```

**Comment by munificent** [comment]
I just added an in-progress proposal for this:

https://github.com/dart-lang/language/commit/329f626a9bae65585065471d1cc59e236d7cf58b

Thanks again for the excellent suggestion @alorenzen!

**Comment by Levi-Lesches** [comment]
Looks great! 

I know this is a separate issue. but would it be possible to extend this to optional parameters as well? 

```dart
final padding = isMobile ? null : EdgeInsets.all(6);
return Container(
  padding: ?padding,
  child: Text("This may or may not have padding"),
);
```

**Comment by lrhn** [comment]
Argument elements is a different issue. It will likely need/want more than just null-aware arguments. At least `if`-arguments to. 
And then there is the complication of omitting a non-trailing positional argument.

**Comment by mit-mit** [comment]
All the planned work in https://github.com/dart-lang/language/issues/323 is now complete, and this is on schedule to launch in the next stable release 🚀
